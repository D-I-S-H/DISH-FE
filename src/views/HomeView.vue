<script setup>
import { ref, toRaw, computed } from 'vue';
import axios from 'axios';
import { useToast } from 'vue-toast-notification';
import 'vue-toast-notification/dist/theme-sugar.css';
import MenuItemComponent from '../components/menuItemComponent.vue';

const API_URL = import.meta.env.VITE_APP_API_URL;
const $toast = useToast();

const menuItems = ref([]); // All menu items fetched from the API
const locations = ref([]); // Locations
const times = ref([]); // Times

const selectedLocation = ref(localStorage.getItem('selectedLocation') || locations.value[0]);
const selectedTime = ref(null); // Selected time

// Fetch locations from the API
axios.get(`${API_URL}/locations`)
  .then(response => {
    locations.value = response.data;
  })
  .catch(error => {
    console.error('Error fetching locations:', error);
    $toast.error('Error fetching locations');
  });

// Selected location
const changeLocation = (location) => {
  selectedLocation.value = location;
  localStorage.setItem('selectedLocation', location);

  // Fetch menu items for the selected location
  axios.get(`${API_URL}/menu?location=${location}`)
    .then(response => {
      menuItems.value = response.data;

      // Set times from response data and default selectedTime
      times.value = response.data.reduce((acc, item) => {
        if (!acc.includes(item.time)) {
          acc.push(item.time);
        }
        return acc;
      }, []);
      
      // Set selectedTime to the first time in the array if available
      if (times.value.length > 0) {
        selectedTime.value = times.value[0];
      }
    })
    .catch(error => {
      console.error('Error fetching menu items:', error);
      $toast.error('Error fetching menu items');
    });
}

// Fetch menu items for the initially selected location
axios.get(`${API_URL}/menu?location=${selectedLocation.value}`)
  .then(response => {
    menuItems.value = response.data;

    // Set times from response data
    times.value = response.data.reduce((acc, item) => {
      if (!acc.includes(item.time)) {
        acc.push(item.time);
      }
      return acc;
    }, []);
    
    // Set selectedTime to the first time in the array if available
    if (times.value.length > 0) {
      selectedTime.value = times.value[0];
    }
  })
  .catch(error => {
    console.error('Error fetching menu items:', error);
    $toast.error('Error fetching menu items');
  });

// Change selected time
const changeTime = (time) => {
  selectedTime.value = time;
};

// Filter menu items
const filteredMenuItems = computed(() => {
  return menuItems.value.filter(item => item.time === selectedTime.value);
});
</script>

<template>
  <div class="container mt-5">
    <div class="row">
      <div class="col-lg-8 mb-3">
        <div class="card mb-3">
          <div class="card-body">
            <div class="d-flex flex-wrap gap-2">
              <button 
                v-for="time in times" 
                :key="time" 
                class="btn"
                :class="{'btn-active': time === selectedTime}"
                @click="changeTime(time)"
              >
                {{ time }}
              </button>
            </div>
          </div>
        </div>
        <div class="card">
          <div class="card-body">
            <h5 class="card-title">Menu</h5>
            <p class="card-text">Select a menu item to view more details.</p>
            <div class="row row-cols-1 row-cols-md-2 g-4">
              <div 
                v-for="menuItem in filteredMenuItems" 
                :key="menuItem.id" 
                class="col"
              >
                <MenuItemComponent :menuItem="menuItem" />
              </div>
            </div>
          </div>
        </div>
      </div>
      
      <div class="col-lg-4 mb-3">
        <div class="card">
          <div class="card-body">
            <h5 class="card-title">Location</h5>
            <ul class="list-group">
              <li 
                v-for="location in locations" 
                :key="location" 
                :class="['list-group-item', 'button-location', {'active': selectedLocation === location}]"
                @click="changeLocation(location)"
              >
                {{ location }}
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
export default {
  name: 'HomeView',
  setup() {
    return { 
      locations, 
      selectedLocation, 
      changeLocation, 
      menuItems, 
      times, 
      selectedTime, 
      changeTime, 
      filteredMenuItems 
    };
  }
};
</script>
