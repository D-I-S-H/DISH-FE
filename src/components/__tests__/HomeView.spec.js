import { mount } from '@vue/test-utils';
import { describe, it, expect, beforeEach, vi, beforeAll } from 'vitest';
import HomeView from "../../views/HomeView.vue";
import axios from 'axios';

vi.mock('axios');

// Mock the environment variable for the API URL
beforeAll(() => {
  process.env.VITE_APP_API_URL = 'http://localhost:8080';
});

describe('HomeView.vue', () => {
  let wrapper;

  beforeEach(async () => {
    axios.get.mockClear(); // Clear any previous mock data

    // Mock localStorage to set a default selected location for testing
    vi.spyOn(localStorage, 'getItem').mockImplementation((key) => {
      if (key === 'selectedLocation') return 'Location 1';
      return null;
    });

    // Mock API responses with a detailed menu item
    axios.get.mockImplementation((url) => {
      if (url.includes('/locations')) {
        return Promise.resolve({ data: ['Location 1', 'Location 2'] });
      }
      if (url.includes('/menu?location=Location%201')) {
        return Promise.resolve({
          data: [
            {
              id: 1,
              name: "Blueberry Coffee Muffin",
              ingredients: [
                "White Cake Mix",
                "Water",
                "Blueberries",
                "Brown Sugar (sugar, cane syrups)",
                "All Purpose Flour",
                "Cinnamon"
              ],
              portion: "1 each",
              description: "Baked Blueberries inside a Fluffy Coffee Cake Muffin",
              nutrients: [
                { name: "Calories", value: "140", uom: "kcal" },
                { name: "Protein (g)", value: "2", uom: "g" },
                { name: "Total Carbohydrates (g)", value: "27", uom: "g" },
                { name: "Sugar (g)", value: "17", uom: "g" },
                { name: "Total Fat (g)", value: "3", uom: "g" },
                { name: "Saturated Fat (g)", value: "1.5", uom: "g" },
                { name: "Cholesterol (mg)", value: "0", uom: "mg" },
                { name: "Dietary Fiber (g)", value: "0", uom: "g" },
                { name: "Sodium (mg)", value: "210", uom: "mg" },
                { name: "Potassium (mg)", value: "30", uom: "mg" },
                { name: "Calcium (mg)", value: "20", uom: "mg" },
                { name: "Iron (mg)", value: "0.6", uom: "mg" },
                { name: "Trans Fat (g)", value: "-", uom: "g" },
                { name: "Vitamin D (IU)", value: "0", uom: "IU" },
                { name: "Vitamin C (mg)", value: "0", uom: "mg" },
                { name: "Calories From Fat", value: "25", uom: "kcal" },
                { name: "Vitamin A (RE)", value: "0+", uom: "RE" },
                { name: "Saturated Fat + Trans Fat (g)", value: "0+", uom: "g" }
              ],
              calories: 140,
              time: "Breakfast",
              location: "McNair",
              allergens: ["Gluten", "Milk", "Wheat", "Egg", "MSG*", "Soy"],
              labels: ["Vegetarian", "How Good Friendly"]
            }
          ]
        });
      }
      return Promise.reject(new Error('Unexpected URL'));
    });

    wrapper = mount(HomeView);
    await wrapper.vm.$nextTick(); // Wait for the component to update after mounting
  });

  it('renders the component correctly', () => {
    expect(wrapper.exists()).toBe(true);
  });

  it('calls the API to get the locations', () => {
    expect(axios.get).toHaveBeenCalledWith('http://localhost:8080/locations');
  });

  // Select the first location by default
  it('displays the first location by default', () => {
    expect(wrapper.vm.selectedLocation).toBe('Location 1');
  });

  // Check if menu was called
  it('calls the API to get the menu for the selected location', () => {
    expect(axios.get).toHaveBeenCalledWith('http://localhost:8080/menu?location=Location 1');
  });



});
