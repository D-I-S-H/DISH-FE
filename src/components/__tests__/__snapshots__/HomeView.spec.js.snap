// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HelloWorld > renders properly 1`] = `
"<div class="container mt-5">
  <div class="card">
    <div class="card-body">
      <h5 class="card-title">Home</h5>
      <p class="card-text">lipsum</p>
    </div>
  </div>
</div>"
`;

exports[`Render home page > renders properly 1`] = `
"<div class="container mt-5">
  <div class="card">
    <div class="card-body">
      <h5 class="card-title">Home</h5>
      <p class="card-text">lipsum</p>
    </div>
  </div>
</div>"
`;

exports[`Render test > renders properly 1`] = `
"<div class="container mt-5">
  <div class="card">
    <div class="card-body">
      <h5 class="card-title">Home</h5>
      <p class="card-text">lipsum</p>
    </div>
  </div>
</div>"
`;
